core/src/cluster_info.rs

```diff
@@ -1098,60 +1104,138 @@ impl ClusterInfo {
+                            if !ret {
+                                inc_new_counter_error!(
+                                    "cluster_info-gossip_pull_response_verify_fail",
+                                    1
+                                );
+                            }
+                            ret
+                        });
+                        Self::handle_pull_response(me, &from, data);
+                    }
+                    Protocol::PushMessage(from, mut data) => {
+                        data.retain(|v| {
+                            let ret = v.verify();
+                            if !ret {
+                                inc_new_counter_error!(
+                                    "cluster_info-gossip_push_msg_verify_fail",
+                                    1
+                                );
+                            }
+                            ret
+                        });
+                        let _ignore_disconnect = response_sender
+                            .send(Self::handle_push_message(me, &from, data, stakes));
+                    }
+                    Protocol::PruneMessage(from, data) => {
+                        if data.verify() {
+                            inc_new_counter_debug!("cluster_info-prune_message", 1);
+                            inc_new_counter_debug!(
+                                "cluster_info-prune_message-size",
+                                data.prunes.len()
+                            );
+                            match me.write().unwrap().gossip.process_prune_msg(
+                                &from,
+                                &data.destination,
+                                &data.prunes,
+                                data.wallclock,
+                                timestamp(),
+                            ) {
+                                Err(CrdsGossipError::PruneMessageTimeout) => {
+                                    inc_new_counter_debug!("cluster_info-prune_message_timeout", 1)
+                                }
+                                Err(CrdsGossipError::BadPruneDestination) => {
+                                    inc_new_counter_debug!("cluster_info-bad_prune_destination", 1)
+                                }
+                                _ => (),
+                            }
+                        } else {
+                            inc_new_counter_debug!("cluster_info-gossip_prune_msg_verify_fail", 1);
+                        }
+                    }
+                    _ => {
+                        let _ignore_disconnect = response_sender
+                            .send(Self::handle_repair(me, &from_addr, blocktree, request));
+                    }
+                })
+        });
+        // process the collected pulls together
+        let _ignore_disconnect =
+            response_sender.send(Self::handle_pull_requests(me, gossip_pull_data));
```

Comment:
@aeyakovenko I updated this gossip thread to immediately respond to a message instead of collecting the responses and sending them together in one large batch.